src/cli.js

```diff
@@ -117,40 +119,91 @@ function Validator() {
 
     const results = Object.fromEntries(filenames.map((key) => [key, null]));
     for (const [filename] of Object.entries(results)) {
-      results[filename] = await validateFile(filename, config, cache);
-
-      if (results[filename].valid === false && config.format === "text") {
-        logErrors(filename, results[filename].errors);
+      results[filename] = await validateFile(filename, config, plugins, cache);
+
+      for (const plugin of plugins) {
+        const message = plugin.logSingleResult(
+          results[filename],
+          filename,
+          config.format,
+        );
+        if (message != null) {
+          logger.log(message);
+          break;
+        }
       }
-      // else: silence is golden
 
       cache.resetCounters();
     }
 
-    if (config.format === "json") {
-      resultsToJson(results);
+    for (const plugin of plugins) {
+      const message = plugin.logAllResults(results, config.format);
+      if (message != null) {
+        logger.log(message);
+        break;
+      }
     }
 
     return resultsToStatusCode(results, config.ignoreErrors);
   };
 }
 
 async function cli(config) {
+  // ** TODO: move all this plugin bootstrapping stuff
+
+  // TODO: expand config file format to allow the user to supply an array of plugins here
+  /*
+  TODO: there's another chicken and egg problem here
+  You need to load plugins (including user plugins!) to validate the config file
+  but you have to parse the config file to load the user plugins
+
+  How are you going to unpick this one?
```

Comment:
I think the way to do this is:

1. Load the config and validate it against the config schema only
2. Load the the plugins (user and core)
3. Validate the config properties that are dependent on which plugins are loaded
4. Parse arguments

That couples loading config and plugins into a single messy "bootstrap" process with multiple responsibilities, but I think it validates my opinion that `INVALID_CONFIG_OR_PLUGIN` is one error code.

Reply:
Accepting `--plugins` on the command line would add more chickens and more eggs to this problem because you then have to be able to load plugins after argument parsing (which in principle partly defines what arguments can be parsed).
Lets not do this, at least to start with. Accepting plugins from the config file only is enough complexity.